### PR TITLE
[Instance Assignment] Add implementations for instance assignment

### PR DESCRIPTION
**Pull request by Jackie-Jiang**
Generate InstancePartitions based on InstanceConfigs and InstanceAssignmentConfig
Generate default InstanceAssignmentConfig from TableConfig for backward-compatibility

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceReplicaPartitionSelector.java [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        int instanceIdInPool = 0;
+        for (int instanceIdInReplica = 0; instanceIdInReplica < numInstancesPerReplica; instanceIdInReplica++) {
+          for (int replicaId : replicaIdsInPool) {
+            replicaIdToInstancesMap[replicaId][instanceIdInReplica] =
+                instanceConfigsInPool.get(instanceIdInPool++).getInstanceName();
+          }
+        }
+      }
+
+      // Assign instances within a replica to one partition if not configured
```
Comment (snleee) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceReplicaPartitionSelector.java]: I am a bit confused on the term that we are using.

`replica group - a set of instances that contains all segments for a partition or table`

`replica - we used to use this term for indicating replication factor for a segment (e.g. 3 replicas)`

Is this still correct in your comments? What do you mean by `instances within a replica`?

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceTagPoolSelector.java [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        poolsToSelect = new ArrayList<>(numPoolsToSelect);
+        List<Integer> poolsInCluster = new ArrayList<>(pools);
+        for (int i = 0; i < numPoolsToSelect; i++) {
+          poolsToSelect.add(poolsInCluster.get((_tableNameHash + i) % numPools));
+        }
+      }
+
+      // Keep the pools selected
+      LOGGER.info("Selecting pools: {} for table: {}", poolsToSelect, _tableNameWithType);
+      pools.retainAll(poolsToSelect);
```
Comment (snleee) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceTagPoolSelector.java]: Why do we need to modify `pools` here? it seems that it's a local variable and it's not used after here?

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceConstraintApplier.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+/**
+ * The instance constraint applier is responsible for filtering out unqualified instances and sorts instances for picking
+ * priority.
+ */
+public interface InstanceConstraintApplier {
+
+  /**
+   * Applies the instance constraint to the given map from pool to instance configs.
+   */
+  Map<Integer, List<InstanceConfig>> applyConstraint(Map<Integer, List<InstanceConfig>> poolToInstanceConfigsMap);
```
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceConstraintApplier.java]: Done.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceReplicaPartitionSelector.java (hunk 1 above)
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceReplicaPartitionSelector.java]: The replica here refers to the table level replica. Added javadoc for the class

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceTagPoolSelector.java (hunk 2 above)
Comment (Jackie-Jiang) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceTagPoolSelector.java]: `pools` is the keySet() of the `poolToInstanceConfigsMap`, so any change to `pools` will reflect to the map.

Comment (snleee) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceTagPoolSelector.java]: ah i see :)

**Pull request merged**